src/charm.py

```diff
@@ -1,117 +1,296 @@
-        """Define and start a workload using the Pebble API.
+        self.certificates = TLSCertificatesRequiresV3(self, TLS_CERT)
+        self.framework.observe(
+            self.on.certificates_relation_created, self._on_certificates_relation_created
+        )
+        self.framework.observe(
+            self.on.certificates_relation_joined, self._on_certificates_relation_joined
+        )
+        self.framework.observe(
+            self.on.certificates_relation_broken, self._on_certificates_relation_broken
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_expiring
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated, self._on_certificate_invalidated
+        )
+        self.framework.observe(self.on.get_certificate_action, self._on_get_certificate_action)
+        self.framework.observe(
+            self.certificates.on.all_certificates_invalidated,
+            self._on_all_certificates_invalidated,
+        )
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def get_hostname(self) -> str:
+        """Get a list containing all ingress hostnames.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Returns:
+            A list containing service and additional hostnames
+        """
+        # The relation will always exist when this method is called
+        service_hostname = cast(str, get_config(self, "service-hostname"))
+        return service_hostname
+
+    def _are_relations_ready(self, event: Union[EventBase, None]) -> bool:
+        """Check if required relations are ready.
 
         Args:
-            event: event triggering the handler.
+            event: The event that fires this method.
+
+        Returns:
+            Whether required relations are ready and execution should continue.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            if event:
+                event.defer()
```

Comment:
Have you considered not deferring here?

Reply:
If we don't defer here then we'll probably need to do it in the `_reconcile` method of the charm, as this hook is only called once when TLS relation is established to create the CSR and store the resulting certificate and key

I'd prefer to keep `event.defer()` here since it makes sense to me to run this only once.

Reply:
what do you think?

Reply:
Won't the handler be triggered again on `certificates_relation_created`?

Reply:
I think the workflow for both hooks is a bit different, on `certificates_relation_joined` we create the CSR and request a certificate from the certificate provider, while on `certificates_relation_created` we generate a secret and a private key and store them as juju secrets

Reply:
Not sure why it's done this way but it's the same in nginx-ingress-integrator I believe.

Reply:
I think using defer in a method which is not the event handler is not compliant with our team standard (it makes also reasoning harder):

https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619

>The only charm methods and functions that accept EventBase instances as arguments should be the event handlers themselves. In other words, every other method should not be interacting with Events. This means, among other things, that only event handlers (the methods on which Framework.observe is called) can defer events. This helps keep the execution flow simple.


 I believe the code base of nginx-ingress-integrator is older than the standard, so you may want to reconsider writing it differently.

Reply:
Make sense, I removed the `event.defer()` from methods that are not event handlers. I still keep defer in certificate related handlers since I think it makes sense to defer there. I'm marking this and other related conversations resolved and we can move the discussion internally if needed.